src/workload.rs

```diff
@@ -438,15 +438,20 @@ impl WorkloadStore {
     }
 
     fn find_upstream(&self, addr: SocketAddr, hbone_port: u16) -> Option<Upstream> {
-        if let Some(upstream) = self.vips.get(&addr) {
+        if let Some(wl_vips) = self.vips.get(&addr) {
             // Randomly pick an upstream
             // TODO: do this more efficiently, and not just randomly
-            let us: &Upstream = upstream.iter().choose(&mut rand::thread_rng()).unwrap();
-            // TODO: avoid clone
-            let mut us: Upstream = us.clone();
-            Self::set_gateway_address(&mut us, hbone_port);
-            debug!("found upstream from VIP: {}", us);
-            return Some(us);
+            let (workload_ip, target_sock_addr) =
+                wl_vips.iter().choose(&mut rand::thread_rng()).unwrap();
+            if let Some(wl) = self.workloads.get(workload_ip) {
+                let mut us = Upstream {
+                    workload: wl.clone(),
+                    port: target_sock_addr.port(),
```

Comment:
why do we need to store the whole target_sock_addr when we only use the port? Can the `vips` map just be SocketAddr -> SocketAddr (service:serviceport -> workload:workloadport)?

Reply:
Hi @howardjohn, sure, how about define the `vips` like below?
```
vips: HashMap<SocketAddr, HashSet<(serviceport, IpAddr)>>
```
As you know, we just want to use service port and workload ip to construct the Upstream. Is that okay?

Reply:
I think you could just turn the tuple into a SocketAddr but yeah basically the same

Reply:
can we make sure we are not going to lose any information if use SocketAddr to take place of Workload?

Reply:
@KfreeZ according to the workload definition, it's enough for us to identify a workload.

Reply:
> @KfreeZ according to the workload definition, it's enough for us to identify a workload.

suppose if we want to use other information of a workload, let's say waypoint address, we have to look it up again in the workloads Hashmap, correct?

Reply:
Yes, it is

Reply:
> 

cool, thanks for the response, no more question.